### ISSUE TITLE
OpenAPI Generated Type Names

### ISSUE THREAD

**Issue by c-reeder** [goadesign/goa#3169]
The names for the types of the Request bodies seem to be solely generated off of the names of the goa method.
Ex. If I have `	Method("create", func() {`, then it will yield `CreateRequestBody`.

The problem with this is that there can be many duplicates.

If I have 5 different services and each one has a `create` method like above, then I will end up with the unfortunately named `CreateRequestBody`, `CreateRequestBody2`, `CreateRequestBody3`, `CreateRequestBody4`, and `CreateRequestBody5` and there is no way to avoid that.

There are a few solutions that come to mind:

- Prefix the name of the service in front of the generated name. Ex. a "create" method inside the "Users" service would yield `UsersCreateRequestBody` or something similar
- Simply make the name of the goa type for that payload the same as the name that is output in the OpenAPI. Thus giving you a way to set the name from code.
- The `TypeName()` function in the DSL currently only affects the generated go type names. This could be altered to produce the same names in the OpenAPI output.
- Some other new DSL function that specifically serves the purpose of letting you set the name of the generated OpenAPI generated type name (whether in an inline defined payload or a separately defined `Type()`.

I'm sure there maybe other similarly helpful solutions, but I think that the main goal here is that we provide:

-  _Some_ better default system for selecting names that makes it much less common for numbers to need to be appended to the end of those generated OpenAPI type names
-  OR some way to override the generated OpenAPI type names
- Or both

Thanks for reading and I'm helping to provide any further information that would enrich this description! :)

**Comment by raphael** [comment]
Makes sense, a few thoughts:
* We need to consider backwards compatibility so whatever solution gets adopted here needs to be "opt-in" so the default does not change.
* There already exists a few ways to modify the generated OpenAPI spec via `Meta`. It would probably be a natural extension to the DSL to add the ability to configure the generated OpenAPI type name there. I'd be happy to merge a PR that did that.
* Piggybacking on `TypeName()` is not guaranteed to always work since different services may define types with identical names that way.

So TL; DR it would make sense to add e.g. `openapi:typename` to the list of special keys to `Meta`. This key would only be valid on attribute expressions (user type definitions are a special kind of such expression) and would override the default type name generated by the OpenAPI codegen.

**Comment by c-reeder** [comment]
@raphael , Thanks for the response!

I'm with you on all points here.

I think the `Meta()` solution would be great if you think it's something that would be straightforward to do! :)

**Comment by c-reeder** [comment]
@raphael , Also just for a bit of extra enrichment here, I notice that the "issue" only seems to occur with the v3 OpenAPI output. Unless I'm mistaken, v2 seems to actually already do what I was suggesting in my first bullet point above:
> Prefix the name of the service in front of the generated name. Ex. a "create" method inside the "Users" service would yield UsersCreateRequestBody or something similar

I'm not as familiar with the rules/customs of OpenAPI, but I just wanted to make sure that it was intended to be implemented this way in v2 but _not_ in v3. 
Thus, if I have a "Users" service, with a "Create" method inside it, v2 gives me:
`                    $ref: '#/definitions/UsersCreateRequestBody'
`
but v3 just gives me:
`                            $ref: '#/components/schemas/CreateRequestBody'`
Is this really the expected behavior? It seems like maybe it was intended that these match. Again, I might have just missed some obvious design reason why one includes the prefix and the other doesn't. 

I'm still 100% on board with the `Meta()` solution you proposed above for overriding the default generated OpenApi names, however it _also_ seems like maybe the default for v3 is a little _off_. Thoughts?

**Comment by raphael** [comment]
Good point, IIRC this was intentional to avoid creating long type names. Instead v3 adds the unique prefix when the types differ. So I still think the `Meta` approach is the right one here.
